bundles/org.openhab.ui.cometvisu/src/main/java/org/openhab/ui/cometvisu/internal/backend/rest/FsResource.java

```diff
@@ -0,0 +1,259 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.ui.cometvisu.internal.backend.rest;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.DefaultValue;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Response.Status;
+
+import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
+import org.glassfish.jersey.media.multipart.FormDataParam;
+import org.openhab.core.io.rest.RESTResource;
+import org.openhab.ui.cometvisu.internal.Config;
+import org.openhab.ui.cometvisu.internal.ManagerSettings;
+import org.openhab.ui.cometvisu.internal.backend.model.rest.ReadResponse;
+import org.openhab.ui.cometvisu.internal.util.FileOperationException;
+import org.openhab.ui.cometvisu.internal.util.FsUtil;
+import org.openhab.ui.cometvisu.internal.util.MountedFile;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+
+/**
+ * Filesystem backend for the cometvisu manager.
+ *
+ * @author Tobias Bräutigam - Initial contribution
+ *
+ */
+@Component
+@Path(Config.COMETVISU_BACKEND_ALIAS + "/fs")
+@Api(Config.COMETVISU_BACKEND_ALIAS + "/fs")
+public class FsResource implements RESTResource {
+    private final Logger logger = LoggerFactory.getLogger(FsResource.class);
+
+    @POST
+    @Consumes("text/*")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Create a text file")
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "OK"), @ApiResponse(code = 403, message = "not allowed"),
+            @ApiResponse(code = 406, message = "File already exists") })
+    public Response create(@QueryParam("path") String path, @QueryParam("type") String type,
```

Comment:
Do I read this correctly that any file in any path can be written through this REST resource (as long as the openHAB user has the fs privileges)? Sounds very dangerous to me.

Reply:
The file access is restricted to a subfolder of the CometVisu installation. The paths are treated as relative paths with this folder as root and the normalize method in class MountedFile should make sure that you cannot breakout this path by using something like `../../`